File: FractionalKnapsack.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.Scanner;

public class FractionalKnapsack {
    private static class Item {
        double value;
        double weight;
        double ratio;

        public Item(double value, double weight) {
            this.value = value;
            this.weight = weight;
            this.ratio = value / weight;
        }
    }

    public static double getMaxValue(double[] values, double[] weights, double capacity) {
        int n = values.length;
        Item[] items = new Item[n];
        for (int i = 0; i < n; i++) {
            items[i] = new Item(values[i], weights[i]);
        }

        Arrays.sort(items, Comparator.comparingDouble((Item i) -> i.ratio).reversed());

        double maxValue = 0;
        for (Item item : items) {
            if (capacity == 0) {
                break;
            }
            double fraction = Math.min(1, capacity / item.weight);
            maxValue += fraction * item.value;
            capacity -= fraction * item.weight;
        }

        return maxValue;
    }


    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);

        System.out.print("Enter count: ");
        int count = Integer.parseInt(input.nextLine());

        double[] values = new double[count];
        double[] weights = new double[count];

        System.out.print("Enter capacity: ");
        double capacity = Double.parseDouble(input.nextLine());

        System.out.println("Enter [value] [weight]");
        for (int i = 0; i < count ; i++) {
            String[] inp = input.nextLine().split(" ");
            values[i] = Double.parseDouble(inp[0]);
            weights[i] = Double.parseDouble(inp[1]);
        }

        double maxValue = FractionalKnapsack.getMaxValue(values, weights, capacity);
        System.out.println("Max value that can be put in the knapsack = " + maxValue);
    }
}

File: fractionalKnapsack.cpp
#include <utility>
#include <iostream>
#include <algorithm>
#include <cstdio>

using namespace std;
double fractionalKnapsack(struct Item arr[], int N, int size);
bool cmp(struct Item a, struct Item b);
struct Item
{
	int value;
	int weight;
	void set(int v, int w)
	{
		value = v;
		weight = w;
	}
};
int main()
{
	int W;
	cout << "Enter your Knapsack maximum weight:" << endl;
	cin >> W;
	int size;
	cout << "Enter number of your items" << endl;
	cin >> size;

	Item* arr = new Item[size];
	cout << "enter your items value and weight" << endl;
	for (int i = 0; i < size; i++)
	{
		int x, y;
		cin >> x;
		cin >> y;
		arr[i].set(x, y);
	}
	cout << "Maximum profit earned = "
		<< fractionalKnapsack(arr, W, size);
	delete arr;
	return 0;
}
double fractionalKnapsack(struct Item arr[],int N, int size)
{
	sort(arr, arr + size, cmp);
	int curWeight = 0;
	double finalvalue = 0.0;

	for (int i = 0; i < size; i++) {
		if (curWeight + arr[i].weight <= N) {
			curWeight += arr[i].weight;
			finalvalue += arr[i].value;
		}
		else {
			int remain = N - curWeight;
			finalvalue += arr[i].value
				* ((double)remain
					/ arr[i].weight);

			break;
		}
	}
	return finalvalue;
}
bool cmp(struct Item a, struct Item b)
{
	double r1 = (double)a.value / a.weight;
	double r2 = (double)b.value / b.weight;
	return r1 > r2;
}
